### PR TITLE
Allow PUT without If-Match in control-http API.  Also allow HEAD.

### LINKED ISSUE #218: namerd: PUT with no If-Match
when writing namerd clients, it's convenient to be able to write a dtab without worrying about concurrent writes.  I.e. "just upload this file".

I think we should support PUT without If-Match to enable this.  I know I had a slightly different take on this originally, but I'm starting to think this is acceptable if sufficiently documented (when namerd has api docs)

### PR DESCRIPTION
**Pull request by adleong**
Fixes #218

### CODE REVIEW

File: namerd/iface/control-http/src/main/scala/io/buoyant/namerd/iface/HttpControlService.scala [hunk 1]
```diff
@@ -120,91 +120,100 @@ class HttpControlService(storage: DtabStore) extends Service[Request, Response]
...
+    }
 
-    // invalid uri
-    case _ => Future.value(Response(Status.NotFound))
-  }
+  private[this] def handleGetDtab(ns: String, req: Request): Future[Response] =
+    getDtab(ns).map {
+      case Some(dtab) =>
+        val rsp = Response()
+        val (contentType, codec) = DtabCodec.accept(req.accept).getOrElse(DtabCodec.default)
```
Comment (olix0r) [inline namerd/iface/control-http/src/main/scala/io/buoyant/namerd/iface/HttpControlService.scala]: i know you didn't write this initially, but I just realized that we're supposed to return a 406 Not Acceptable if we are given an Accept header we can't satisfy.  I think in general, the accept stuff I did is not particularly RFC compliant, so maybe best to revisit in a followup

**Comment by olix0r** [comment]
:+1: looks good!

File: namerd/iface/control-http/src/main/scala/io/buoyant/namerd/iface/HttpControlService.scala [hunk 2]
```diff
@@ -120,91 +120,100 @@ class HttpControlService(storage: DtabStore) extends Service[Request, Response]
...
+        rsp.contentType = contentType
+        rsp.headerMap.add("ETag", versionString(dtab.version))
```
Comment (gtcampbell) [inline namerd/iface/control-http/src/main/scala/io/buoyant/namerd/iface/HttpControlService.scala]: minor: finagle-http appears to have "ETag" as a constant in `Fields`; might as well use that

File: namerd/iface/control-http/src/main/scala/io/buoyant/namerd/iface/HttpControlService.scala [hunk 3]
```diff
@@ -120,91 +120,100 @@ class HttpControlService(storage: DtabStore) extends Service[Request, Response]
...
+
+      case None => Response(Status.NotFound)
+    }
+
+  private[this] def handlePutDtab(ns: String, req: Request): Future[Response] =
+    req.contentType.flatMap(DtabCodec.byContentType) match {
+      case Some(codec) =>
+        codec.read(req.content) match {
+          case Return(dtab) =>
+            req.headerMap.get("If-Match") match {
```
Comment (gtcampbell) [inline namerd/iface/control-http/src/main/scala/io/buoyant/namerd/iface/HttpControlService.scala]: can use the constant in finagle-http here as well

File: namerd/iface/control-http/src/main/scala/io/buoyant/namerd/iface/HttpControlService.scala [hunk 4]
```diff
@@ -120,91 +120,100 @@ class HttpControlService(storage: DtabStore) extends Service[Request, Response]
...
+    }
+
+  private[this] def handleHeadDtab(ns: String, req: Request): Future[Response] =
+    getDtab(ns).map {
+      case Some(dtab) =>
+        val rsp = Response()
+        rsp.headerMap.add("ETag", versionString(dtab.version))
+        rsp
+
+      case None => Response(Status.NotFound)
```
Comment (gtcampbell) [inline namerd/iface/control-http/src/main/scala/io/buoyant/namerd/iface/HttpControlService.scala]: tioli: get and head have a bunch of shared logic (look up dtab, 404 if not found, add etag to response) that could be refactored into a shared method

**Comment by gtcampbell** [comment]
:+1: this looks fine to me, although it would probably be good to switch to the finagle-http constants as opposed to the hard-coded strings here

**Pull request merged**